src/nsrt_learning/nsrt_learning_main.py

```diff
@@ -224,3 +256,88 @@ def _learn_pnad_samplers(pnads: List[PartialNSRTAndDatastore],
+    """Function to check whether a given set of operators and predicates
+    preserves a sigle training trajectory."""
+    init_atoms = utils.abstract(ll_traj.states[0], init_preds)
+    objects = set(ll_traj.states[0])
+    ground_nsrts, _ = task_plan_grounding(init_atoms, objects, strips_ops,
+                                          option_specs)
+    heuristic = utils.create_task_planning_heuristic(
+        CFG.sesame_task_planning_heuristic, init_atoms, traj_goal,
+        ground_nsrts, init_preds, objects)
+
+    def _check_goal(state: Tuple[FrozenSet[GroundAtom], int]) -> bool:
+        return traj_goal.issubset(state[0])
+
+    def _get_successor_with_correct_option(
+        searchnode_state: Tuple[FrozenSet[GroundAtom], int]
+    ) -> Iterator[Tuple[_GroundNSRT, Tuple[FrozenSet[GroundAtom], int],
+                        float]]:
+        state = searchnode_state[0]
+        idx_into_traj = searchnode_state[1]
+
+        if idx_into_traj <= len(ll_traj.actions) - 1:
+            assert ll_traj.actions[idx_into_traj].has_option()
+            gt_option = ll_traj.actions[idx_into_traj].get_option()
+            expected_next_hl_state = hl_traj[idx_into_traj + 1]
+
+            for applicable_nsrt in utils.get_applicable_operators(
+                    ground_nsrts, state):
+                # NOTE: we check that the ParameterizedOptions are equal before
+                # attempting to ground because otherwise, we might
+                # get a parameter mismatch and trigger an AssertionError
+                # during grounding.
+
+                if applicable_nsrt.option == gt_option.parent:
+                    if applicable_nsrt.option_objs == gt_option.objects:
+                        next_hl_state = utils.apply_operator(
+                            applicable_nsrt, set(state))
+                        # Here, we check whether all atoms that differ
+                        # between next_hl_state and state are part of
+                        # the operator's side predicates. If so, this nsrt
+                        # can be applied from this state!
+                        exp_state_matches = next_hl_state.issubset(
+                            expected_next_hl_state)
+
+                        if exp_state_matches:
+                            # The returned cost is uniform because we don't
+                            # actually care about finding the shortest path;
+                            # just one that matches!
+                            yield (applicable_nsrt, (frozenset(next_hl_state),
+                                                     idx_into_traj + 1), 1.0)
+
+    init_atoms_frozen = frozenset(init_atoms)
+    init_searchnode_state = (init_atoms_frozen, 0)
+    state_seq, _ = utils.run_gbfs(
+        init_searchnode_state, _check_goal, _get_successor_with_correct_option,
+        lambda searchnode_state: heuristic(searchnode_state[0]))
+
+    return _check_goal(state_seq[-1])
```

Comment:
What would be a situation where this `_check_goal` does not return True, since you're using `_check_goal` as the termination condition of the `run_gbfs` call above? Oh, are you worried about the case that `run_gbfs` hits its max_expansions or max_evals limit? That makes sense. Did that come up in practice?

Reply:
well also remember that `run_gbfs` might not be able to succeed if we've sidelined too aggressively or something (it'll just hit a bunch of dead-ends). So I think this check is necessary